r/src/array__to_vector.cpp

```diff
@@ -0,0 +1,827 @@
+    IntegerVector data(no_init(n));
+    auto dict_array = static_cast<DictionaryArray*>(arrays[0].get());
+    auto dict = dict_array->dictionary();
+    auto indices = dict_array->indices();
+    switch (indices->type_id()) {
+      case Type::UINT8:
+      case Type::INT8:
+      case Type::UINT16:
+      case Type::INT16:
+      case Type::INT32:
+        break;
+      default:
+        stop("Cannot convert Dictionary Array of type `%s` to R",
+             dict_array->type()->ToString());
+    }
+
+    if (dict->type_id() != Type::STRING) {
+      stop("Cannot convert Dictionary Array of type `%s` to R",
+           dict_array->type()->ToString());
+    }
+    bool ordered = dict_array->dict_type()->ordered();
+
+    data.attr("levels") = ArrayVector_To_Vector<Converter_String>(dict->length(), {dict});
+    if (ordered) {
+      data.attr("class") = CharacterVector::create("ordered", "factor");
+    } else {
+      data.attr("class") = "factor";
+    }
+    return data;
+  }
+
+  static Status Ingest(SEXP data_, const std::shared_ptr<arrow::Array>& array,
+                       R_xlen_t start, R_xlen_t n) {
+    DictionaryArray* dict_array = static_cast<DictionaryArray*>(array.get());
+    auto indices = dict_array->indices();
+    switch (indices->type_id()) {
+      case Type::UINT8:
+        return Ingest_Impl<arrow::UInt8Type>(data_, array, start, n);
+      case Type::INT8:
+        return Ingest_Impl<arrow::Int8Type>(data_, array, start, n);
+      case Type::UINT16:
+        return Ingest_Impl<arrow::UInt16Type>(data_, array, start, n);
+      case Type::INT16:
+        return Ingest_Impl<arrow::Int16Type>(data_, array, start, n);
+      case Type::INT32:
+        return Ingest_Impl<arrow::Int32Type>(data_, array, start, n);
+      default:
+        break;
+    }
+    return Status::OK();
+  }
+
+  template <typename Type>
+  static Status Ingest_Impl(SEXP data_, const std::shared_ptr<arrow::Array>& array,
+                            R_xlen_t start, R_xlen_t n) {
+    IntegerVector_ data(data_);
+
+    DictionaryArray* dict_array = static_cast<DictionaryArray*>(array.get());
+    using value_type = typename arrow::TypeTraits<Type>::ArrayType::value_type;
+    auto null_count = array->null_count();
+
+    if (n == null_count) {
+      std::fill_n(data.begin() + start, n, NA_INTEGER);
+    } else {
+      std::shared_ptr<Array> indices = dict_array->indices();
+      auto p_array = indices->data()->GetValues<value_type>(1);
+      if (!p_array) {
+        return Status::Invalid("invalid data buffer");
+      }
+      if (array->null_count()) {
+        arrow::internal::BitmapReader bitmap_reader(indices->null_bitmap()->data(),
+                                                    indices->offset(), n);
+        for (size_t i = 0; i < n; i++, bitmap_reader.Next(), ++p_array) {
+          data[start + i] =
+              bitmap_reader.IsNotSet() ? NA_INTEGER : (static_cast<int>(*p_array) + 1);
```

Comment:
I've added a comment about the +1, and moved to a lambda. 

I don't think it's worth it having a bounds check for `INT32_MAX` because it is fairly unlikely that there will be that many levels in a factor.